include/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.hpp

```diff
@@ -228,6 +230,7 @@ class Propagator
 
    private:
     Shared<CoordinatesBroker> coordinatesBrokerSPtr_ = std::make_shared<CoordinatesBroker>();
+    StateBuilder stateBuilder_;
```

Comment:
I love the fact that we can use `StateBuilder` now.

My only comment is whether it's probably best to just create the integration `StateBuilder` whenever we need, rather than having it as a property of the `Propagator`.

So far it is used 3 times:
- in `calculateStateAt`
- in `calculateStateToCondition`
- in `caclulateStatesAt`

I'd say than rather than just creating it as a property. We just initialise at those 3 instances. 

```cpp
const StateBuilder integrationStateBuilder = StateBuilder(Propagator::IntegrationFrameSPTr, coordinatesBrokerSPtr_);
```

If we don't want to have this instantiation repeated 3 times, we can have a private fun that generates it and then we just call it whenever we need:

```cpp
const StateBuilder integrationStateBuilder = this->generateIntegrationStateBuilder();
```

**Why?**

Because at the moment there is some logic to keep the `Propagator` in "synch" with the dynamics and other external elements. For instance, whenever we clear dynamics:

```cpp
void Propagator::clearDynamics()
{
    this->dynamicsContexts_.clear();
    this->coordinatesBrokerSPtr_ = std::make_shared<CoordinatesBroker>();
}
```

We would probably need to take care here of updating the `stateBuilder_` so that we make sure it points to the proper coordinates broker:

```Cpp
void Propagator::clearDynamics()
{
    this->dynamicsContexts_.clear();
    this->coordinatesBrokerSPtr_ = std::make_shared<CoordinatesBroker>();

      // This line below;
    this->stateBuilder_ = {Propagator::IntegrationFrameSPtr, this->coordinatesBrokerSPtr_};}
```

Whilst **this is still ok for now**, as the "keeping-in-synch" logic is not very complex. **It could get cumbersome in the future** (for example, we now no longer have a hardcode integration frame, but a variable one, or another function that causes the coordinate broker to change it's pointer), so I think it's safer to just the integration stateBuilder on a needly basis:
- We ensure the `stateBuilder` is always "in-synch"
- One less thing to worry about in the functions that keep stuff "in-synch" (e.g. `clearDynamics`)
- Creating a `stateBuilder` is not expensive

...and you wouldn't have to modify/update your tests, as this is all happening behind the scenes : )

Reply:
Since the coordinates broker is a shared pointer. And the statebuilder has a shared pointer of the broker, if we update the broker anywhere, it gets updated in the state builder as well. However you're right in the fact that it's used in a pretty trivial way right now so there isn't a strong case for having it be a member variable. I'm fine with both approaches as they achieve the same outcome with about the same performance and readibility.

Reply:
That's a great point Pau. I waffled between the two approaches and actually had it the other way at first. 

I think you're right that it's safer to just create it on-demand for now. It can easily become a property in the future if it makes sense to do that.

Reply:
This is a snippet that validates that updating the broker would update it within the StateBuilder as well as they are shared pointers:
```
// Online C++ compiler to run C++ program online
#include <iostream>
#include <memory>

struct Cat
{
    std::string name;
    
    Cat(std::string aName): name(aName) {};
};

struct Zoo
{
    std::shared_ptr<Cat> cat1;
    
    Zoo(const std::shared_ptr<Cat>& aCat): cat1(aCat) {};
};

int main() {
    // Write C++ code here
    std::shared_ptr<Cat> toto = std::make_shared<Cat>(Cat("toto"));
    
    std::cout << "toto name: " << toto->name << std::endl;
    Zoo zoo(toto);
    std::cout << "cat1 name: " << zoo.cat1->name << std::endl;
    
    toto->name = "not toto";
    std::cout << "toto name: " << toto->name << std::endl;
    std::cout << "cat1 name: " << zoo.cat1->name << std::endl;

    return 0;
}
```

Reply:
Thanks for the snippet Vishwa 🙏 . It does confirm the intuition. I've still just made them local properties for now, since it seems that we all agree it won't be too impactful and it's easy to go the other way in the future if needed.

Reply:
For sure yeah, not advocating in either direction. I just wanted to confirm that's how shared pointers worked lol :P